File: leetcode/tree/337. House Robber III/js/main.js
const rob = function (root) {
  const dfs = root => {
    if (root == null) {
      return [0, 0];
    }

    let left = dfs(root.left);
    let right = dfs(root.right);

    let res = [0, 0];

    res[0] = Math.max(left[0], left[1]) + Math.max(right[0], right[1]);
    res[1] = root.val + left[0] + right[0];

    return res;
  };

  let ans = dfs(root, 0);

  return Math.max(ans[0], ans[1]);
};

File: leetcode/tree/337. House Robber III/py/main.py
# Time O(N)
# Space O(N)
class Solution:
    def rob(self, root):
        if not root:
          return 0

        def isLeaf(node):
          return node.left == None and node.right == None
        
        def dfs(node):
          if node == None:
            return [0, 0]

          if isLeaf(node):
            return [node.val, 0]

          l = dfs(node.left)
          r = dfs(node.right)
          
          # [current sum, max reachable sum] 
          return [node.val + l[1] + r[1], max(l) + max(r)]  
            
        return max(dfs(root))
